Add ability to map root of subgroup

dbudworth opened the pull request:
There was a bug in the initial implementation of subgroups where mapping the root of a subgroup ended up in adding the "add trailing slash" logic.

This code basically changes it so if you do subGroup.Handle("/",...) then we just use the subgroups original path.

This change maps:
router.NewGroup("/foo").GET("/") -> /foo (no trailing slash added)

before this change it did:
/foo (with add trailing slash enabled)

dimfeld commented:
My only concern here is that this makes it impossible to actually define a '/' on a subgroup when you do want the trailing slash added. 

Perhaps a better solution would be modify `checkPath` to allow passing in an empty string, changing it to check `if len(path) > 0 && path[0] != '/'`. Then, to get the behavior you desire, you could just call `group.GET("", handler)`, while invalid calls such as `group.GET("bar", handler)` are still disallowed. What do you think?

dbudworth commented:
yep, that's probably a better idea 

We are saying that to map the subgroup itself, we do:
NewGroup("/foo").GET("", handler)
and to get traling slash logic, we do:
NewGroup("/foo").GET("/", handler)

That gives you the option to do either with the only downside being that, I assume, the most common usage will be to pass an empty string which feels weird for an API.

That said, I don't really have an alternative to offer.

dimfeld commented:
Yeah, that's what I'm thinking. I agree it feels a little weird, but I also couldn't come up with a better alternative.

dbudworth commented:
ok, updated to reflect changes and added test cases for all 3 versions
NewGroup("/foo").GET("") + GET /foo results in 200
NewGroup("/foo").GET("/") + GET /foo results in 301 -> /foo/
NewGroup("/foo").GET("/") + GET /foo/ results in 200

dimfeld commented:
Thanks!
